### ISSUE TITLE
Can RunContainer numberOfRuns and ArrayContainer cardinality be smaller?

### ISSUE THREAD

**Issue by richardstartin** [RoaringBitmap/RoaringBitmap#369]
It just occurred to me that there can only be 2048 runs in a `RunContainer` and only 4096 values in an `ArrayContainer` - could `numberOfRuns` and `cardinality` be `short`s? Whenever they are used as array indexes, this will lead to zero extension to a 32 bit integer, which will have a small performance impact. Has this tradeoff been considered before?

**Comment by lemire** [comment]
Once serialized, we do use only 16-bit values to indicate the number of runs and the cardinality (of any/a) container. It may or may not be so tight when implemented in memory. (It is not, actually!)

Are you thinking about saving 2 bytes here and there?

I don't think that the zero extension would prove to be a bottleneck, but it may make the code slightly more difficult to deal with in Java since there are no uint16_t.

**Comment by richardstartin** [comment]
Yes.

> more difficult to deal with in Java since there are no uint16_t.

Well, there's `char`...

**Comment by lemire** [comment]
Yes. You are correct: `char` would work. It would be better than `short` for sure.

It is unclear to me whether there would be any benefit in practice, but I am certainly open to the idea that there might be.

**Comment by Maaartinus** [comment]
> Yes. You are correct: `char` would work. It would be better than `short` for sure.

I'm not that sure. It's surely less error-prone as you can't forget ensuring the zero extension. OTOH I'd bet that `int x = aChar` takes exactly as long as `int x = aShort & 0xFFFF` (I haven't measure it). Semantically, I find `char` used as a number rather confusing (in Java)... but given that you're hardly ever dealing with chars, that's not too bad.
